// react-native-chatgpt/lib/commonjs/constants/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.USER_AGENT = exports.STREAMED_REQUEST_DEFAULT_TIMEOUT = exports.REQUEST_DEFAULT_TIMEOUT = exports.PROMPT_ENDPOINT = exports.LOGIN_SUCCESS_PAGE = exports.LOGIN_PAGE = exports.HOST_URL = exports.CHAT_PAGE = void 0;
var _reactNative = require("react-native");
const HOST_URL = 'https://chat.openai.com';
exports.HOST_URL = HOST_URL;
const CHAT_PAGE = `${HOST_URL}/chat`;
exports.CHAT_PAGE = CHAT_PAGE;
const LOGIN_PAGE = `${HOST_URL}/auth/login`;
exports.LOGIN_PAGE = LOGIN_PAGE;
const PROMPT_ENDPOINT = `${HOST_URL}/backend-api/conversation`;
exports.PROMPT_ENDPOINT = PROMPT_ENDPOINT;
const LOGIN_SUCCESS_PAGE = `${HOST_URL}/api/auth/callback`;

// We have to set a user agent to bypass Google security. Google login is not supported by webviews.
// By setting the user agent we trick Google into thinking we are using the Chrome mobile browser.
// The user agent is obtained from https://user-agents.net/string/mozilla-5-0-iphone-cpu-iphone-os-10-3-like-mac-os-x-applewebkit-602-1-50-khtml-like-gecko-crios-90-0-2924-75-mobile-14e5239e-safari-602-1
// https://github.com/react-native-webview/react-native-webview/issues/162
exports.LOGIN_SUCCESS_PAGE = LOGIN_SUCCESS_PAGE;
const USER_AGENT = _reactNative.Platform.OS === 'ios' ? 'Mozilla/5.0 (iPhone; CPU iPhone OS 10_3 like Mac OS X) AppleWebKit/602.1.50 (KHTML, like Gecko) CriOS/90.0.2924.75 Mobile/14E5239e Safari/602.1' : 'Mozilla/5.0 (Linux; Android 13; SM-N981B Build/TP1A.220624.014) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/108.0.5359.128 Mobile Safari/537.36 [FB_IAB/Orca-Android;FBAV/392.0.0.12.106;]';
exports.USER_AGENT = USER_AGENT;
const REQUEST_DEFAULT_TIMEOUT = 30000;
exports.REQUEST_DEFAULT_TIMEOUT = REQUEST_DEFAULT_TIMEOUT;
const STREAMED_REQUEST_DEFAULT_TIMEOUT = 15000;
exports.STREAMED_REQUEST_DEFAULT_TIMEOUT = STREAMED_REQUEST_DEFAULT_TIMEOUT;
