// react-native-chatgpt/lib/commonjs/hooks/useWebViewAnimation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = useWebViewAnimation;
var _react = require("react");
var _reactNative = require("react-native");
function useWebViewAnimation(_ref) {
  let {
    onAnimationStart,
    onAnimationEnd
  } = _ref;
  const animatedValue = (0, _react.useRef)(new _reactNative.Animated.Value(0));
  const translateY = animatedValue.current.interpolate({
    inputRange: [0, 0.5, 1],
    outputRange: [_reactNative.Dimensions.get('window').height, _reactNative.Dimensions.get('window').height, 0],
    extrapolate: 'clamp'
  });
  const backdropOpacity = animatedValue.current.interpolate({
    inputRange: [0, 0.8, 1],
    outputRange: [0, 0.5, 0.5]
  });
  const animateWebView = mode => {
    onAnimationStart === null || onAnimationStart === void 0 ? void 0 : onAnimationStart(mode);
    _reactNative.Animated.timing(animatedValue.current, {
      toValue: mode === 'show' ? 1 : 0,
      duration: 600,
      useNativeDriver: true
    }).start(() => {
      onAnimationEnd === null || onAnimationEnd === void 0 ? void 0 : onAnimationEnd(mode);
    });
  };
  return {
    animatedStyles: {
      webview: {
        transform: [{
          translateY
        }]
      },
      backdrop: {
        opacity: backdropOpacity
      }
    },
    animateWebView
  };
}
